File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def traverse(root, prev):
            if root is None:
                return True

            if not traverse(root.left, prev):
                return False

            if root.val <= prev[0]:
                return False

            prev[0] = root.val

            return traverse(root.right, prev)

        prev = [float("-inf")]

        return traverse(root, prev)

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool isBST(struct TreeNode* node, long minVal, long maxVal) {
    if (node == NULL) {
        return true;
    }
    if (node->val <= minVal || node->val >= maxVal) {
        return false;
    }
    return isBST(node->left, minVal, node->val) && isBST(node->right, node->val, maxVal);
}

bool isValidBST(struct TreeNode* root) {
    return isBST(root, LONG_MIN, LONG_MAX);
}